pkg/promutil/migrate.go

```diff
@@ -64,63 +63,76 @@ func BuildMetrics(results []model.CloudwatchMetricResult, labelsSnakeCase bool,
 	for _, result := range results {
 		contextLabels := contextToLabels(result.Context, labelsSnakeCase, logger)
 		for _, metric := range result.Data {
-			for _, statistic := range metric.Statistics {
-				var includeTimestamp bool
-				if metric.AddCloudwatchTimestamp != nil {
-					includeTimestamp = *metric.AddCloudwatchTimestamp
-				}
-				exportedDatapoint, timestamp, err := getDatapoint(metric, statistic)
-				if err != nil {
-					return nil, nil, err
-				}
-				if exportedDatapoint == nil && (metric.AddCloudwatchTimestamp == nil || !*metric.AddCloudwatchTimestamp) {
-					exportedDatapoint = aws.Float64(math.NaN())
-					includeTimestamp = false
-					if *metric.NilToZero {
-						exportedDatapoint = aws.Float64(0)
+			var statisticsRepresentedInMetric []string
+			includeTimestamp := false
+			if metric.MetricConfig.AddCloudwatchTimestamp != nil {
+				includeTimestamp = *metric.MetricConfig.AddCloudwatchTimestamp
+			}
+			// A result from GetMetricData is for a single statistic noted on the result while a
+			// result from GetMetricStatistics is for all statistics configured for the metric
+			if metric.GetMetricDataResult != nil {
+				statisticsRepresentedInMetric = []string{metric.GetMetricDataResult.Statistic}
+			} else {
+				statisticsRepresentedInMetric = metric.MetricConfig.Statistics
+			}
+			for _, statistic := range statisticsRepresentedInMetric {
+				var exportedDatapoint float64
+				var timestamp time.Time
+				if metric.GetMetricDataResult != nil {
+					exportedDatapoint = metric.GetMetricDataResult.Datapoint
```

Comment:
Yeah a comment here makes sense the content will be slightly different though. The way I refactored it, it's not possible at all for the value to be `nil` as it's no longer a pointer. It is possible that it is `NaN` though if the GetMetricData API returns a NaN value to us. I couldn't find if this was possible or not but if I search for `NaN` in the sdk I see that their have support for `NaN` so it _seems_ possible.

Reply:
There's a bug in recent versions where absent metrics are always returned at `0` instead of `nil`/`NaN`, because `MetricDataResult.Datapoint` wasn't a pointer and defaulted everything to `0`.

I attempted to address this in #1288 by making `MetricDataResult.Datapoint` a pointer and ensuring that `nil` is passed through for metrics with no values, though I see that this change makes `nil` impossible in either layer as now neither are pointers. In practice AWS does not return `NaN` datapoints for missing metrics, they return `Values: []` instead, and `pkg/clients/cloudwatch/*.toMetricDataResult()` has to check for this.

You could use `NaN` as a marker value for absent metrics if you want to avoid pointers, but `toMetricDataResult()` would need to be updated to explicitly populate that `NaN` instead of using the default value of `0` in `MetricDataResult.Datapoint`

Reply:
Agree that adding a marker in `toMetricDataResult()` is a more appropriate fix than the one proposed in https://github.com/nerdswords/yet-another-cloudwatch-exporter/pull/1288. If you're all ok with it, I think the problem could be addressed in this PR while we're refactoring models.

Reply:
This is awesome because it's the exact reason I wanted to refactor the shape of this struct. There's too much reliance on implicit behavior rather than explicit, in this case there was a reliance on the fact that the datapoint was default initialized to `nil` rather than explicitly being marked as "not present". My plan is to split the new struct I made `GetMetricDataResult` in to two,

```go
type GetMetricDataProcessingParams struct {
	// GetMetricData Query MetricQueryID used for results mapping
	MetricQueryID string
	// Indicates if this result was successfully mapped to a query result, only used during metric gathering
	// and does not need to be respected afterward
	MappedToAQueryResult bool
}

type GetMetricDataResult struct {
	Statistic    string
	HadDatapoint bool
	Datapoint    float64
	Timestamp    time.Time
}
```

The eventually I would drop `GetMetricDataProcessingParams` from this struct and hide that complexity behind the new GetMetricDataProcessor abstraction (https://github.com/kgeckhart/yet-another-cloudwatch-exporter/pull/2). 

I'm choosing to be super explicit with a `HadDatapoint bool` vs using `NaN` in the Datapoint so we can tell if the `Timestamp` value is usable for `AddCloudwatchTimestamp` (https://github.com/kgeckhart/yet-another-cloudwatch-exporter/blob/47cbd27471262253a832b826b1ae0c8eef216529/pkg/promutil/migrate.go#L87-L91).

Reply:
@nhinds I rebased my PR on top of yours because I didn't want to lose the great work you did to add tests where they did not exist previously. Thank you so much for doing this!

I did notice one odd thing when I pulled them over where for one of the new tests you added this metric, https://github.com/nerdswords/yet-another-cloudwatch-exporter/pull/1288/files#diff-74544d5174e5eca3f8554da703851b2282e736b61cce5530e02d405b31ebbd94R453-R468, is dropped after the call to `BuildMetrics`. After I finished the rebase I had to update the test to include it in the results which I think is valid because `BuildMetrics` shouldn't be causing any metrics to be dropped. If you would be able to look at it that would be helpful.

Reply:
> I was hoping it would leave the changesets associated with you but it looks like it pulled your name but not your github info

...so, it turns out I committed those with my work details, which aren't tied to my personal github account. That's my fault, I was going to re-do them if I hadn't closed my PR and didn't realize you would pull them in here. Thanks for picking up the commits, I was considering sending a followup PR to add the tests once this was merged and I'm glad to not have to rewrite them.

> I would happily take my changes and apply them to your branch if you felt comfortable giving me access to push to the branch in your fork or any other option so you don't lose them.

I've updated the branch on my fork with the correct author (https://github.com/nhinds/yet-another-cloudwatch-exporter/tree/allow-nil-nan-metrics) if you wanted to rebase, but feel free to leave them as-is if that's too much effort.

> After I finished the rebase I had to update the test to include it in the results which I think is valid because BuildMetrics shouldn't be causing any metrics to be dropped. If you would be able to look at it that would be helpful.

With the code on my branch, `GetMetricDataPoint` would be `nil` if there were no datapoints returned from CloudWatch. If there are no datapoints, there are also no datapoint timestamps - I think my test data is slightly misleading there, it should have also set `GetMetricDataTimestamps: time.Time{}` rather than `GetMetricDataTimestamps: ts`.

That particular testcase checked the logic where if `AddCloudwatchTimestamp` is `true` and there is no timestamp available from Cloudwatch, the metric is skipped as we don't have a timestamp from cloudwatch.

It looks like the new behaviour in this PR will result in a metric being returned as `NaN` or `0` without a timestamp in this scenario, which I think that will cause Prometheus to interpret the `NaN` or `0` as a datapoint with a timestamp equal to the scrape time. That might cause problems if a future scrape _does_ find data from cloudwatch, as the timestamp from cloudwatch may end up being earlier than the last scrape time. I think Prometheus doesn't like ingesting metrics whose timestamps end up out-of-order.

Personally I do not use `AddCloudwatchTimestamp`, so it will not impact me either way. I don't know if this will be a practical problem for people who _do_ use it, but I would be hesitant to change the behaviour if it would be simple to continue to skip metrics without results when the user has requested timestamps be returned from cloudwatch.

Reply:
If @cristiangreco is okay with it @nhinds what I think would be easiest here is you re-open your PR and it gets merged. I really didn't intend this PR to change any logic I wanted to just move around fields and remove unnecessary pointers. I happened to stubble on to this part which looked weird while you actually fixed it + added tests.

The corner case with `AddCloudwatchTimestamp` is rather annoying but I also agree that dropping the metric is the right thing to do. More recent versions of prometheus support out of order samples but even then the data ends up being incredibly confusing to use. CloudWatch timestamps are always behind by the period amount so a non-CloudWatch timestamp is going be in line with the next CloudWatch sample but not rounded due to scrape time. I'll clean this up on my branch + add some clarifying comments. All of this + the tests will hopefully prevent a future regression like this again.

Reply:
Sorry for the late reply. Yes please, @nhinds can you re-open your PR and tag me for review?

Reply:
Reopened #1288